doc/peapod.8
.TH PEAPOD 8 "May 30 2018" "peapod" ""


.SH NAME

peapod \- EAPOL Proxy Daemon


.SH SYNOPSIS

.TP 7
.B peapod
.B "[\-dtsqCoh]"
.B "[\-vvv]"
.BI "[\-p " pidfile "]"
.BI "[\-c " configfile "]"
.BI "[\-l [" logfile "]]"


.SH DESCRIPTION

.B peapod
is a daemon that proxies
.B "IEEE 802.1X Extensible Authentication Protocol over LAN (EAPOL)"
packets between Ethernet interfaces. It supports a few tricks on a
per\-interface basis, so it may be considered a (highly) rudimentary
general\-purpose transparent bridging firewall/rewriting proxy for EAPOL.

.B EAPOL
is a
.B "port\-based network access control (PNAC)"
mechanism ensuring that only authorized devices are allowed to use a network.
In a nutshell, EAPOL blocks regular network traffic, such as TCP/IP, from
traversing the physical port (e.g. on a switch) to which a client is connected
until the client successfully authenticates and establishes an EAPOL session.

Abilities surpassing those of a simple proxy include:

.TP
.B "EAPOL/EAP classification, filtering, and script execution"
.B "Proxy only certain kinds of packets"
between certain interfaces and
.B "execute user\-defined scripts"
when proxying recognized packet types.
This is supported for the nine EAPOL Packet Types defined by IEEE Std
802.1X\-2010 and the four EAP Codes defined by IETF RFC 2284.

.TP
.B "802.1Q VLAN priority tag handling"
.B "Add, modify, or remove priority tags"
in proxied EAPOL packets. (In fact, other fields besides the Priority Code
Point field in the 802.1Q VLAN tag may be manipulated.)

.TP
.B "MAC spoofing"
Change interface MAC to a
.BR "user\-defined address" ,
or to the address of an actual supplicant behind the proxy learned during
runtime. This enables the device running
.B peapod
to masquerade as the supplicant and originate what appears to be authorized
network traffic once the supplicant establishes an EAPOL session (as long as
MACsec is not in use).


.SH OPERATION

At startup,
.B peapod
reads its config file to determine the network interfaces on which it should
listen for EAPOL packets and how packets should be received and sent on those
interfaces. \(dqEAPOL packet\(dq in this sense is an Ethernet frame with the
EAPOL EtherType of
.B 0x888e
encapsulating either an EAP packet or certain EAPOL control messages.

See
.BR peapod.conf (5)
for config file semantics and syntax.

The remainder of this section provides an overview of
.BR peapod 's
control flow once it begins proxying packets proper, and is intended to assist
in the creation of config files. This can be divided into
.B ingress
and
.B egress
phases.

.SS "ingress phase"
.B peapod
waits until an EAPOL packet arrives on any configured interface, which then
becomes the
.BR "ingress interface" .

The following is then performed:

.RS
.IP "1. " 3
If the packet is the first packet to be received on the ingress interface, and
.I another
interface is configured with
.RB \(dq set\-mac\-from
.IR ingress\-interface\-name \(dq:
.RS 4
.IP \(bu 2
Change the
.I other
interface's MAC address to that of the packet's sender.
.IP \(bu 2
Drop the packet.
.RE
.IP "2. " 3
Determine the type of the packet.
.RS 4
.IP \(bu 2
Check for
.B "ingress scripts"
and/or
.B "ingress filters"
(scripts/filters configured as
.B "ingress stanza options"
on the ingress interface) that match the packet type.
.RE
.IP "3. " 3
Execute any matching ingress script.
.IP "4. " 3
Apply any matching ingress filter (i.e. drop the packet).
.RE

.PP
Dropping a packet in this phase means immediately discarding it and returning
to the beginning of the ingress phase.

If the packet was not dropped,
.B peapod
moves on to the egress phase.

.SS "egress phase"
The received packet is proxied to all configured interfaces except for the
ingress interface, which then become the
.BR "egress interfaces" .

For each egress interface, the following is then performed:

.RS
.IP "1. " 3
Make a copy of the packet to be used with the current egress interface.
.IP "2. " 3
If any existing 802.1Q VLAN tag in the packet should be manipulated or removed,
or one should be added (as configured in a
.B dot1q
option on the current egress interface):
.RS 4
.IP \(bu 2
Make the necessary changes to the copy.
.RE
.IP "3. " 3
Determine the type of the packet.
.RS 4
.IP \(bu 2
Check for
.B "egress filters"
and/or
.B "egress scripts"
(filters/scripts configured as
.B "egress stanza options"
on the current egress interface) that match the packet type.
.RE
.IP "4. " 3
Apply any matching egress filter (i.e. drop the packet).
.IP "5. " 3
Execute any matching egress script.
.IP "6. " 3
Send the (copy of the) packet on the current egress interface.

.PP
Dropping a packet in this phase means discarding the copy and moving on to the
next egress interface.

At the end of the egress phase, the packet may have been proxied to one or more
egress interfaces, and
.B peapod
returns to the beginning of the ingress phase to listen for more packets.


.SH EXAMPLES

Example config files and scripts are included in
.BR peapod 's
shared resources in
.IR "/usr/share/peapod/examples" ,
as well as in the
.I "doc/examples"
subdirectory of the program sources.

See
.BR peapod.conf (5)
for config file semantics and syntax.


.SH OPTIONS

All short options except
.B "\-v"
also have a long form listed. Mandatory arguments are mandatory for both forms.

.TP
.BR "\-d" , " \-\-daemon"
Run as a daemon. Logging to the console
.RB ( stdout
and
.BR stderr )
is disabled. Sets
.B \-s
and causes a PID file to be used.

.TP
.BR "\-p " \f[I]pidfile\f[R], " \f[B]\-\-pid " \f[I]pidfile
Specify a different PID file than
.IR /var/run/peapod.pid ,
the default. Sets
.BR \-d .

.TP
.BR "\-c " \f[I]configfile\f[R], " \f[B]\-\-config " \f[I]configfile
Specify a different config file than
.IR /etc/peapod.conf ,
the default.

A config file is required. See
.BR peapod.conf (5)
for more details.

.TP
.BR "\-t" , " \-\-test"
Test the config file and exit.

.TP
.BR "\-l " [\f[I]logfile\f[R]], " \-\-log " [\f[I]logfile\f[R]]
Enable logging to a log file. Optionally, specify a different log file than
.IR /var/log/peapod.log ,
the default.

.TP
.BR "\-s" , " \-\-syslog"
Enable logging to syslog. Set automatically by
.B \-d
and
.BR \-p .

.TP
.B "\-v"
Increase log verbosity. Can be specified up to three times. Overridden by the
.B verbosity
option in the config file.

With
.BR "\-v" ,
informational messages are also logged.
.br
With
.BR "\-vv" ,
so are debug messages.
.br
With
.BR "\-vvv" ,
so are low\-level debug messages such as Ethernet frame hexdumps. (As these last
are extremely voluble, they are logged only to the console and/or to a log file,
and never to syslog.)

.TP
.BR "\-q" , " \-\-quiet\-script"
Treat script execution notices as informational, so that they are
logged only if at least one
.B "\-v"
was provided.

.TP
.BR "\-C" , " \-\-color"
Colorize logging output to console.

.TP
.BR "\-o", " \-\-oneshot"
Do not restart the proxy after certain errors occur, such as a configured
interface going down unexpectedly.

The default error handling behavior once the
proxy is running is to wait ten seconds between unlimited restart attempts.

.TP
.BR "\-h", " \-\-help"
Print a help message to the console.


.SH FILES

.nf
.I /usr/sbin/peapod
.I /etc/peapod.conf
.I /var/log/peapod.log
.I /var/run/peapod.pid
.fi


.SH BUGS

Definitely. For suggestions, bug reports, contributions, pull requests, etc.,
please contact the author via the project page at
<https://github.com/kangtastic/peapod> or via e\-mail.

While not a bug per se, note that
.BR peapod 's
usefulness is greatly limited on MACsec\-enabled networks.


.SH "SEE ALSO"

.BR peapod.conf (5)


.SH AUTHORS

.nf
James Seo (kangtastic) <kangscinate@gmail.com>
.fi


debian/changelog
peapod (0.1.1-1) unstable; urgency=low

  * New upstream release

 -- James Seo <kangscinate@gmail.com>  Thu, 12 Jul 2018 02:13:34 -0700

peapod (0.1.0-1) unstable; urgency=low

  * Initial release

 -- James Seo <kangscinate@gmail.com>  Sat, 07 Jul 2018 13:12:11 -0700


src/peapod.c
/**
 * @file peapod.c
 * @brief EAPOL Proxy Daemon
 */
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <paths.h>
#include <signal.h>
#include <string.h>
#include <unistd.h>
#include "args.h"
#include "daemonize.h"
#include "defaults.h"
#include "log.h"
#include "parser.h"
#include "proxy.h"

static void help_exit(int status);
static void signal_handler(int sig);
int main(int argc, char *argv[]);

/**
 * @name Signal counters
 * @note Global
 * @{
 */
volatile sig_atomic_t sig_hup = 0;
volatile sig_atomic_t sig_int = 0;
volatile sig_atomic_t sig_usr1 = 0;
volatile sig_atomic_t sig_term = 0;
/** @} */

/** @brief Program usage string */
static const char usage[] = {
"\n"
"%s - EAPOL Proxy Daemon\n"
"\n"
"Usage: %s [-dtsqnoh] [-vvv] [-p <pidfile>] [-c <conffile>] [-l [<logfile>]]\n"
"\n"
"Mandatory arguments are mandatory for both forms of an option.\n"
"\n"
"  -d, --daemon         run as daemon (sets -s, disables console output)\n"
"  -p, --pid=PATH       set PID file (sets -d, default: %s)\n"
"\n"
"  -c, --config=PATH    set config file (default: %s)\n"
"  -t, --test           test config file and exit\n"
"\n"
"  -l, --log[=PATH]     output to a log file (default: %s)\n"
"\n"
"  -s, --syslog         output to syslog\n"
"\n"
"  -v                   verbosity of output - can be specified up to 3 times\n"
"                       -v:   additionally output informational messages\n"
"                       -vv:  also output debug messages\n"
"                       -vvv: also output low-level debug messages such as data\n"
"                             structure views and EAPOL packet hexdumps to the\n"
"                             console and/or a log file (but not to syslog)\n"
"\n"
"  -q, --quiet-script   only output script execution notices with at least one -v\n"
"\n"
"  -C, --color          colorize console output\n"
"\n"
"  -o, --oneshot        do not restart proxy after certain errors occur\n"
"\n"
"  -h, --help           print this help and exit\n"
"\n"
"%s version %s\n"
};

/** @brief An environment containing only @p PATH */
static char *clean_environ[] = { "PATH=" _PATH_STDPATH,	NULL };

/**
 * @brief Environment variables
 * @note Global, provided by compiler
 */
extern char **environ;
extern struct args_t args;

/**
 * @brief Interface list
 * @note Global
 */
struct iface_t *ifaces;

/**
 * @brief Print usage information to @p stderr and exit
 * @param status The exit status to pass to the @p exit(2) system call
 */
static void help_exit(int status)
{
	cerr(usage, PEAPOD_PROGRAM, PEAPOD_PROGRAM, PEAPOD_PID_PATH,
	     PEAPOD_CONF_PATH, PEAPOD_LOG_PATH, PEAPOD_PROGRAM, PEAPOD_VERSION);
	exit(status);
}

/**
 * @brief Signal handler
 *
 * Increment signal counters upon receiving a signal. If more than one @p SIGINT
 * or @p SIGTERM has been received without being acted upon, abort the program.
 *
 * @param sig The signal that was received
 */
static void signal_handler(int sig)
{
	switch (sig) {
	case SIGHUP:
		++sig_hup;
		break;
	case SIGINT:
		if (++sig_int > 1)
			abort();
		break;
	case SIGUSR1:
		++sig_usr1;
		break;
	case SIGTERM:
		if (++sig_term > 1)
			abort();
		break;
	default:
		break;
	}
}

/**
 * @brief Close all open file descriptors except @p stdin, @p stdout, and
 *        @p stderr
 * @return 0 if successful, or -1 if unsuccessful
 */
int peapod_close_fds(void)
{
	for (int i = getdtablesize() - 1; i > 2; --i) {
		if (close(i) == -1 && errno != EBADF) {
			ecrit("couldn't close file descriptor %d: %s", i);
			return -1;
		}
	}
	return 0;
}

/**
 * @brief Redirect @p stdin, @p stdout, and @p stderr to @p /dev/null
 * @return 0 if successful, or -1 if unsuccessful
 */
int peapod_redir_stdfds(void)
{
	int devnull = open("/dev/null", O_RDONLY, 0);
	if (devnull == -1) {
		ecrit("cannot open /dev/null readonly: %s");
		return -1;
	}
	if (dup2(devnull, STDIN_FILENO) == -1) {
		ecrit("cannot redirect stdin: %s");
		return -1;
	}
	close(devnull);

	devnull = open("/dev/null", O_WRONLY, 0);
	if (devnull == -1) {
		ecrit("cannot open /dev/null writeonly: %s");
		return -1;
	}
	if (dup2(devnull, STDOUT_FILENO) == -1) {
		ecrit("cannot redirect stdout: %s");
		return -1;
	}
	close(devnull);

	devnull = open("/dev/null", O_RDWR, 0);
	if (devnull == -1) {
		ecrit("cannot open /dev/null readwrite: %s");
		return -1;
	}
	if (dup2(devnull, STDERR_FILENO) == -1) {
		ecrit("cannot redirect stderr: %s");
		return -1;
	}
	close(devnull);

	return 0;
}

/**
 * @brief Main function
 * @param argc The number of command-line arguments
 * @param argv An array of command-line arguments (i.e. C strings)
 * @return 0 by default
 */
int main(int argc, char *argv[])
{
	struct sigaction sa;
	memset(&sa, 0, sizeof(struct sigaction));

	sigset_t sigmask;
	sigemptyset(&sigmask);
	sigaddset(&sigmask, SIGHUP);
	sigaddset(&sigmask, SIGINT);
	sigaddset(&sigmask, SIGUSR1);
	sigaddset(&sigmask, SIGTERM);
	if (sigprocmask(SIG_BLOCK, &sigmask, NULL) == -1) {
		cerr("cannot set signal mask");
		exit(EXIT_FAILURE);
	}

	sa.sa_handler = signal_handler;
	sigemptyset(&sa.sa_mask);
	sa.sa_flags = 0;
	sigaction(SIGHUP, &sa, NULL);
	sigaction(SIGINT, &sa, NULL);
	sigaction(SIGUSR1, &sa, NULL);
	sigaction(SIGTERM, &sa, NULL);

	if (args_get(argc, argv) == -1)
		help_exit(EXIT_FAILURE);

	if (args.help == 1)
		help_exit(EXIT_SUCCESS);

	/* Don't log to file or syslog if all we're doing is testing config */
	if (args.test == 1) {
		args.level = LOG_WARNING;
		uint8_t dummy;
		printf("testing config file\n");
		ifaces = parse_config(args.conffile, &dummy);
		printf("config file at '%s' seems valid, exiting\n",
		       args.conffile);
		exit(EXIT_SUCCESS);
	}

	if (log_init() == -1)
		help_exit(EXIT_FAILURE);

	ifaces = parse_config(args.conffile, &args.level);

	uid_t uid = getuid();

	info("running as user %d", (int)uid);

	/* We'll probably faceplant soon, but perhaps we have privileges */
	if (uid != 0)
		warning("not running as root");

	/* Sanitize the environment */
	environ = clean_environ;

	if (args.daemon == 1)
		daemonize(args.pidfile);

	debuglow("printing interface list");
	parser_print_ifaces(ifaces);

	proxy(ifaces);
}
